components/style/properties/longhand/font.mako.rs

```diff
@@ -1006,3 +1006,865 @@ ${helpers.single_keyword("font-variant-position",
         Err(())
     }
 </%helpers:longhand>
+
+<%helpers:longhand name="font-variant-alternates" products="none" animatable="False" boxed="True"
+                   spec="https://drafts.csswg.org/css-fonts-3/#font-variant-alternates-prop">
+    use std::fmt;
+    use style_traits::ToCss;
+    use values::HasViewportPercentage;
+    use values::computed::ComputedValueAsSpecified;
+
+    impl ComputedValueAsSpecified for SpecifiedValue {}
+    no_viewport_percentage!(SpecifiedValue);
+
+    pub mod computed_value {
+        pub use super::SpecifiedValue as T;
+    }
+
+    #[derive(Debug, Clone, PartialEq)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub struct Style {
+        pub historical_forms: bool,
+        pub styleset: Option<Vec<String>>,
+        pub character_variant: Option<Vec<String>>,
+        pub swash: Option<String>,
+        pub ornaments: Option<String>,
+        pub annotation: Option<String>,
+    }
+
+    #[derive(Debug, Clone, PartialEq)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub enum SpecifiedValue {
+        Normal,
+        Style(Style),
+    }
+
+    impl ToCss for SpecifiedValue {
+        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            match *self {
+                SpecifiedValue::Normal => dest.write_str("normal"),
+                SpecifiedValue::Style(ref style) => {
+
+                    let mut is_empty = true;
+
+                    if style.historical_forms {
+                        try!(dest.write_str("historical-forms"));
+                        is_empty = false;
+                    }
+
+                    if let Some(ref args) = style.styleset {
+                        if !is_empty {
+                            try!(dest.write_str(" "));
+                        }
+
+                        try!(dest.write_str("styleset("));
+
+                        let mut first = true;
+                        for arg in args {
+                            if !first {
+                                try!(dest.write_str(", "));
+                            }
+                            first = false;
+
+                            try!(dest.write_str(arg));
+                        }
+
+                        try!(dest.write_str(")"));
+
+                        is_empty = false;
+                    }
+
+                    if let Some(ref args) = style.character_variant {
+                        if !is_empty {
+                            try!(dest.write_str(" "));
+                        }
+
+                        try!(dest.write_str("character-variant("));
+
+                        let mut first = true;
```

Comment:
Since everything's either an `Option<_>` or a `Vec<_>`, and we're doing first argument and `Some(_)` checks everywhere, I think we could go for a macro.